### ISSUE TITLE
the event request_success should pass request and response to the listener

### ISSUE THREAD

**Issue by tobeczm** [locustio/locust#1413]
### Is your feature request related to a problem? Please describe.
I want to fetch the headers of request or response when a request is successful.

### Describe the solution you'd like
```python
@events.request_success.add_listener
def my_success_handler(request_type, name, response_time, response_length, request, response, **kw):
    print("Successfully made a request to: %s" % name)
```

**Comment by heyman** [comment]
That wouldn't work when using Locust to load test non-HTTP stuff. I'd recommend to either create a custom client, or wrap the client calls instead.

**Comment by tobeczm** [comment]
> That wouldn't work when using Locust to load test non-HTTP stuff. I'd recommend to either create a custom client, or wrap the client calls instead.

At least, we could put them into the optional arguments:
```
@events.request_success.add_listener
def my_success_handler(request_type, name, response_time, response_length, **kw):
    print('request:', kw['request'])
    print('response:', kw['response']
    print("Successfully made a request to: %s" % name)
```

**Comment by heyman** [comment]
> At least, we could put them into the optional arguments:

That's a possibility. The only issue I see with that is that people would start relying on their existence and all of a sudden we've crippled support for non-HTTP stuff in practice. Not sure if it would be worth it - maybe?

**Comment by felix-hubert-eai** [comment]
Was also looking for a way to access the request and response in the listener. Any chances this will be possible?

On a similar idea, is it possible to access the `runner` inside a listener? I'm interested in having access to the current `user_count` inside a request_success hook.

**Comment by heyman** [comment]
> On a similar idea, is it possible to access the runner inside a listener? I'm interested in having access to the current user_count inside a request_success hook.

The `runner` is available [through the Environment instance](https://docs.locust.io/en/stable/api.html#locust.env.Environment.runner), and you can grab a reference to it from the [init event](https://docs.locust.io/en/stable/api.html#locust.event.Events.init):

```python
from locust import events

@events.init.add_listener
def _(environment, **kwargs):
    @events.request_success.add_listener
    def _(**kw):
        print("Current user count:", environment.runner.user_count)
```

**Comment by ludeknovy** [comment]
Hi, I support this idea as well, it would be very handy to have request/response at disposal. I would like to be able to access status code for enhanced reporting from listener. Do we have any final decision on this already?

**Comment by cyberw** [comment]
I like the idea and will happily merge it, but someone else needs to write the code and some test cases :)

**Comment by ludeknovy** [comment]
@cyberw btw is it actually a good idea to pass whole request and response from the performance perspective? 

My use case could be easily covered by adding _hostname_ and _status_code_ for now:
```
self.environment.events.request_success.fire(
                    request_type=request_meta["method"],
                    name=request_meta["name"],
                    response_time=request_meta["response_time"],
                    response_length=request_meta["content_size"],
                    request_hostname=request_meta["hostname"],
                    response_status_code=response.status_code
                )
```

**Comment by cyberw** [comment]
I think that could be fine, but it is not related to this ticket. There is another one about that I think...

**Comment by cyberw** [comment]
(ignore my first, now deleted, response. I was on mobile and confused on which ticket this was)

Passing the response body (which might not even be read into Python unless the user asks for it) would be a big no-no for performance, but passing the response _object_  (same as we are passing to the `with` block when catch_response=True) shouldnt be too bad. If we can get the request object as well that would be great. 

I’d also like to pass the User object. 

Tbh, I’m not super worried about other protocols not suppling these values, but we should definitely mention it in our ”new client”-example in the docs.

**Comment by cyberw** [comment]
Fixed in 1.5.0 #1750 (except for the request)
